### PR TITLE
APP-4932 - Refactor Validation to avoid infinte loops in mobile

### PR DESCRIPTION
**Pull request by symphony-adnane**
## Description

In mobile the order of blur and change are not aligned with the desktop browser
This made validation to enter in an infinite loop for dropdown component as it triggers both events on item selection

## JIRA ticket

[JIRA_Ticket](https://perzoinc.atlassian.net/browse/APP-4932 )

### CODE REVIEW

File: packages/components/spec/components/validation/Validation.spec.tsx [hunk 1]
```diff
@@ -3,48 +3,55 @@ import * as React from 'react';
...
+      /** TESTS SHOULD NOT TARGET INTERNALS */
       const validate = jest.spyOn(Validation.prototype, 'updateState');
       const wrapper = shallow(
         <Validation validator={Validators.Required} errorMessage={'Required'}>
           <TextField />
         </Validation>
       );
       expect(wrapper.length).toEqual(1);
-      expect(wrapper.find('.tk-validation--errors').exists()).toBeFalsy();
+      expect(wrapper.find('.tk-validation--error').exists()).toBeFalsy();
```
Comment (AliciaSymphony) [inline packages/components/spec/components/validation/Validation.spec.tsx:24]: 😱

File: packages/components/spec/components/validation/Validation.spec.tsx [hunk 2]
```diff
@@ -3,48 +3,55 @@ import * as React from 'react';
...
-          <TextField onChange={zone.onChange} />
+          <TextField onChange={change} />
         </Validation>
       );
       const mockEvent = { target: { value: 'This is just for test' } };
       wrapper.find('TextField').simulate('change', mockEvent);
       expect(change).toHaveBeenCalledWith(mockEvent);
-      await validate;
-      expect(valChange).toHaveBeenCalledWith(true, null);
+      await currentEventLoopEnd();
```
Comment (AliciaSymphony) [inline packages/components/spec/components/validation/Validation.spec.tsx:51]: Curiosity: What is the rational of the number of `await currentEventLoopEnd()` in this workaround?

Comment (symphony-adnane) [reply packages/components/spec/components/validation/Validation.spec.tsx:51]: It's related to debouncing the update method
I think the more it debounces the more we need to wait for it to settle :/

File: packages/components/src/components/validation/Validation.tsx [hunk 3]
```diff
@@ -51,6 +50,8 @@ class Validation extends React.Component<
     lastAction: null,
   };
 
+  private debouncedUpdateState = debounce(this.updateState);
```
Comment (symphony-annaLlorens) [inline packages/components/src/components/validation/Validation.tsx:53]: 👍

File: packages/components/src/components/validation/Validation.tsx [hunk 4]
```diff
@@ -61,11 +62,11 @@ class Validation extends React.Component<
     // call updateState whenever value or errorsChildMap change
     if (
       this.state.lastAction !== ACTION.RESET &&
-      this.state !== ACTION.ON_INIT &&
+      this.state.lastAction !== ACTION.ON_INIT &&
       (prevState.lastValue !== this.state.lastValue ||
         prevState.errorsChildMap !== this.state.errorsChildMap)
     ) {
-      this.updateState(this.state.lastValue);
+      this.debouncedUpdateState(this.state.lastValue);
```
Comment (symphony-annaLlorens) [inline packages/components/src/components/validation/Validation.tsx:69]: When this be triggered?

**Review by symphony-annaLlorens — APPROVED** [review]

File: packages/components/src/components/validation/Validation.tsx (hunk 4 above)
Comment (symphony-adnane) [reply packages/components/src/components/validation/Validation.tsx:69]: this runs at a state change

File: packages/components/src/components/validation/Validation.tsx [hunk 5]
```diff
@@ -170,7 +172,8 @@ class Validation extends React.Component<
     if (this.props.errors) {
       errors = this.props.errors;
     }
-    this.setState({ isValid: valid, errors, lastValue: value });
+    // don't change the value here otherwise there are risks we get into an infinite loop
+    this.setState({ isValid: valid, errors });
```
Comment (AliciaSymphony) [inline packages/components/src/components/validation/Validation.tsx:176]: so this was the culprit

**Review by AliciaSymphony — APPROVED** [review]

File: packages/components/src/components/validation/Validation.tsx (hunk 5 above)
Comment (symphony-jean-michael) [reply packages/components/src/components/validation/Validation.tsx:176]: Well done @symphony-adnane 👍

**Pull request merged**